### PR TITLE
[WebAssembly] Use 64-bit table when targeting wasm64

### PR DESCRIPTION
**Pull request by sbc100**
See https://github.com/WebAssembly/memory64/issues/51

### CODE REVIEW

**Comment by sbc100** [comment]
I'm planning on writing a table64 lowering pass in binaryen before landing this.

**Comment by aheejin** [comment]
If we don't do `i32.wrap_i64` for table index anymore, can our `call_indirect` take an i64?

**Comment by sbc100** [comment]
> If we don't do `i32.wrap_i64` for table index anymore, can our `call_indirect` take an i64?

Yes with a 64-bit table all indexes must be 64-bit too, including those passed to `call_indirect`

**Comment by aheejin** [comment]
It still only takes a `table32_op` operand.. Am I missing something?
https://github.com/llvm/llvm-project/blob/23f8fac745bdde70ed4f9c585d19c4913734f1b8/llvm/lib/Target/WebAssembly/WebAssemblyInstrCall.td#L63

**Comment by sbc100** [comment]
> It still only takes a `table32_op` operand.. Am I missing something?
> 
> https://github.com/llvm/llvm-project/blob/23f8fac745bdde70ed4f9c585d19c4913734f1b8/llvm/lib/Target/WebAssembly/WebAssemblyInstrCall.td#L61-L67

I guess that code might need updating too.

**Comment by sbc100** [comment]
> > It still only takes a `table32_op` operand.. Am I missing something?
> > https://github.com/llvm/llvm-project/blob/23f8fac745bdde70ed4f9c585d19c4913734f1b8/llvm/lib/Target/WebAssembly/WebAssemblyInstrCall.td#L61-L67
> 
> I guess that code might need updating too.

Actually isn't `table32_op` here referring to the table immediate?  i.e. the index of the table itself, not the index into the table which is taken as an argument?   That fact that its called `$table` (and that the tests pass) makes me think this.

**Comment by aheejin** [comment]
Ah you're right I'm confused. Then I think we should fix the non-isel path too: 
https://github.com/llvm/llvm-project/blob/f608ac261781b7707b2721563765e07c57366619/llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp#L579-L591

Also there will be tests that need fixing.. such as https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/WebAssembly/fast-isel-call-indirect64.ll

Also the change needs to be reflected in the type checker to (not necessarily in this PR):
https://github.com/llvm/llvm-project/blob/cf8b93d8234c13b6a6baf7f8a1e5bdb8c1c8cca1/llvm/lib/Target/WebAssembly/AsmParser/WebAssemblyAsmTypeCheck.cpp#L378-L380

**Comment by sbc100** [comment]
See https://github.com/WebAssembly/binaryen/pull/6595

**Comment by sbc100** [comment]
Updated (and removed) tests.

Updated ISelLowering.

File: lld/wasm/SyntheticSections.cpp [hunk 1]
```diff
@@ -584,12 +584,10 @@ void ElemSection::writeBody() {
   initExpr.Extended = false;
   if (ctx.isPic) {
     initExpr.Inst.Opcode = WASM_OPCODE_GLOBAL_GET;
-    initExpr.Inst.Value.Global =
-        (config->is64.value_or(false) ? WasmSym::tableBase32
-                                      : WasmSym::tableBase)
-            ->getGlobalIndex();
```
Comment (aheejin) [inline lld/wasm/SyntheticSections.cpp:590]: Why is this removed?

Comment (sbc100) [reply lld/wasm/SyntheticSections.cpp:590]: Because we the table is now 64-bit or 32-bit according to whether wasm64 is in use, it now always matches the type of the `__table_base` global.  Prior to this we couldn't use `__table_base` in the case of wasm64 because `__table_base` is 64-bit but the table was only 32-bit, so we needed to use `__tabe_base32`.

This is another advantage to using table64.

**Review by aheejin — APPROVED** [review]

**Comment by sbc100** [comment]
Now waiting on https://github.com/emscripten-core/emscripten/pull/21950 to land
